tempesta_fw/http_sticky.c

```diff
@@ -0,0 +1,740 @@
+	state->fragnum = 0;
+
+	return 0;
+}
+
+#define S_CRLF		"\r\n"
+#define S_CRLFCRLF	"\r\n\r\n"
+#define S_HTTP		"http://"
+
+#define S_302		"HTTP/1.1 302 Found"
+#define S_502		"HTTP/1.1 502 Bad Gateway"
+
+#define S_F_HOST		"Host: "
+#define S_F_DATE		"Date: "
+#define S_F_CONTENT_LENGTH	"Content-Length: "
+#define S_F_LOCATION		"Location: "
+#define S_F_CONNECTION		"Connection: "
+#define S_F_SET_COOKIE		"Set-Cookie: "
+
+#define S_V_DATE		"Sun, 06 Nov 1994 08:49:37 GMT"
+#define S_V_CONTENT_LENGTH	"9999"
+
+#define SLEN(s)		(sizeof(s) - 1)
+
+#define	S_302_FIXLEN							\
+	SLEN(S_302)							\
+	+ SLEN(S_CRLF)							\
+	+ SLEN(S_F_DATE) + SLEN(S_V_DATE)				\
+	+ SLEN(S_CRLF)							\
+	+ SLEN(S_F_CONTENT_LENGTH) + SLEN(S_V_CONTENT_LENGTH)		\
+	+ SLEN(S_CRLF)							\
+	+ SLEN(S_F_LOCATION) + SLEN(S_HTTP)				\
+	+ SLEN(S_CRLF)							\
+	+ SLEN(S_F_SET_COOKIE)						\
+	+ SLEN(S_CRLFCRLF)
+
+static size_t
+tfw_http_prep_date(u_char *buf)
+{
+	/* TODO: put in the current date stamp */
+	memcpy(buf, S_V_DATE, SLEN(S_V_DATE));
+	return SLEN(S_V_DATE);
+}
+
+static size_t
+tfw_http_prep_sticky(u_char *buf, u_char *value, size_t len)
+{
+	u_char *ptr = buf;
+
+	while (len--) {
+		snprintf(ptr, 100, "%02x", (*value++ & 0x0FF));
+		ptr += 2;
+	}
+	return (ptr - buf);
+}
+
+/*
+ * Send an HTTP 302 response to the client. The response redirects
+ * the client to the same URI as the original request, but includes
+ * 'Set-Cookie:' header field that sets Tempesta sticky cookie.
+ */
+static int
+tfw_http_send_302(TfwHttpMsg *hm)
+{
+	u_char buf[256];
+	TfwMsg *msg;
+	TfwStr *chunk;
+	TfwHttpResp *resp;
+	tfw_mastate_t state = { 0 };
+	TfwHttpReq *req = (TfwHttpReq *)hm;
+	size_t len, data_len = S_302_FIXLEN;
+	TfwConnection *conn = hm->conn;
+	TfwClient *client = (TfwClient *)conn->peer;
+
+	if (client->cookie.len == 0) {
+		return -1;
+	}
+	if ((resp = (TfwHttpResp *) tfw_http_msg_alloc(Conn_Srv)) == NULL) {
+		return -1;
+	}
+	msg = (TfwMsg *)resp;
+	data_len += req->uri_path.len
+		    + tfw_cfg_sticky.name.len + 1 + sizeof(client->hmac) * 2;
+	if (req->host.len) {
+		data_len += req->host.len;
+	} else {
+		data_len += hm->h_tbl->tbl[TFW_HTTP_HDR_HOST].field.len;
+	}
+
+	if (tfw_http_msg_setup(&state, msg, data_len) != 0) {
+		tfw_http_msg_free((TfwHttpMsg *)resp);
+		return -1;
+	}
+
+	tfw_http_msg_add_data(&state, msg, S_302, SLEN(S_302));
+	tfw_http_msg_add_data(&state, msg, S_CRLF, SLEN(S_CRLF));
+
+	tfw_http_msg_add_data(&state, msg, S_F_DATE, SLEN(S_F_DATE));
+	len = tfw_http_prep_date(buf);
+	tfw_http_msg_add_data(&state, msg, buf, len);
+	tfw_http_msg_add_data(&state, msg, S_CRLF, SLEN(S_CRLF));
+
+	tfw_http_msg_add_data(&state, msg, S_F_CONTENT_LENGTH,
+					   SLEN(S_F_CONTENT_LENGTH));
+	tfw_http_msg_add_data(&state, msg, "0", 1);
+	tfw_http_msg_add_data(&state, msg, S_CRLF, SLEN(S_CRLF));
+
+	tfw_http_msg_add_data(&state, msg, S_F_LOCATION, SLEN(S_F_LOCATION));
+	tfw_http_msg_add_data(&state, msg, S_HTTP, SLEN(S_HTTP));
```

Comment:
What's the sense to call the function so many times just with static strings as an argument? Why not to call it only once with concatenated static strings?